code/modules/mob/living/simple_animal/bot/turtle/turtle_cartridge.dm

```diff
@@ -0,0 +1,280 @@
+
+/obj/item/disk/turtle_cartridge/proc/format_string(S)
+	for(var/varname in variables)
+		S = replacetext(S, "\[[varname]\]", "[variables[varname]]")
+	return S
+
+/obj/item/disk/turtle_cartridge/proc/process_val(val)
+	if(!isnull(text2num(val)))
+		return text2num(val)
+	if(val in variables)
+		return variables[val]
+	if(text2dir(val)) //direction vars are builtin, but can be overidden, see two lines above
+		return text2dir(val)
+
+	if(is_varname(val))
+		throw_error("UNDEFINED VAR", val)
+	else
+		throw_error("SYNTAX ERROR", val)
+
+/obj/item/disk/turtle_cartridge/proc/process_expression(list/args)
+	if(args.len == 1)
+		return process_val(args[1])
+	else if(args.len == 2)
+		switch(args[1])
+			if("NOT")
+				return !process_val(args[2])
+			if("ABS")
+				return abs(process_val(args[2]))
+			if("FLOOR")
+				return FLOOR(process_val(args[2]), 1)
+			if("CEIL")
+				return CEILING(process_val(args[2]), 1)
+			if("ROUND")
+				return round(process_val(args[2]))
+			if("NEG")
+				return -process_val(args[2])
+			if("SQRT")
+				return sqrt(process_val(args[2]))
+			else
+				throw_error("SYNTAX ERROR")
+	else if(args.len == 3)
+		switch(args[2])
+			if("ADD")
+				return process_val(args[1]) + process_val(args[3])
+			if("SUB")
+				return process_val(args[1]) - process_val(args[3])
+			if("MUL")
+				return process_val(args[1]) * process_val(args[3])
+			if("DIV")
+				if (process_val(args[3]) == 0) return throw_error("DIV BY 0")
+				return process_val(args[1]) / process_val(args[3])
+			if("MOD")
+				if (process_val(args[3]) == 0) return throw_error("DIV BY 0")
+				return process_val(args[1]) % process_val(args[3])
+			if("POW")
+				return process_val(args[1]) ** process_val(args[3])
+			if("MIN")
+				return min(process_val(args[1]), process_val(args[3]))
+			if("MAX")
+				return max(process_val(args[1]), process_val(args[3]))
+			if("LT")
+				return process_val(args[1]) < process_val(args[3])
+			if("LE")
+				return process_val(args[1]) <= process_val(args[3])
+			if("GT")
+				return process_val(args[1]) > process_val(args[3])
+			if("GE")
+				return process_val(args[1]) >= process_val(args[3])
+			if("EQ")
+				return process_val(args[1]) == process_val(args[3])
+			if("NE")
+				return process_val(args[1]) != process_val(args[3])
+			if("OR")
+				return process_val(args[1]) || process_val(args[3])
+			if("AND")
+				return process_val(args[1]) && process_val(args[3])
+			else
+				throw_error("SYNTAX ERROR")
+
+/obj/item/disk/turtle_cartridge/proc/process_line()
+	if(!lines[line])
+		stop()
+		return
+
+	turtle?.speak("PROCESSING LINE [lines[line]]")
+
+	var/list/arguments = splittext(lines[line], " ")
+	if(!arguments.len) return
+
+	var/command = arguments[1]
+	arguments.Cut(1,2)
+	var/full_input = jointext(arguments, " ")
+
+	var/wait = 1
+
+	if (stack.len)
+		var/datum/stack_item/I = stack[stack.len]
+		if (command == "END")
+			if (istype(I, /datum/stack_item/cmd_while))
+				if(I.skipping)
+					stack.len --
```

Comment:
Does this actually work if you have a space before `--`?

Reply:
Yes but it looks bad so I'll change it